### PR TITLE
fix: Local Credential Hijacking fix

### LINKED ISSUE #476: Local HTTP server credential hijacking
Impact
The default browser-based authentication mechanism leaves users vulnerable to a credential hijacking attack, enabling an attacker to gain unauthorised access to a user's Amazon account.

Description
The default authentication mechanism used by ASK CLI involves the following steps:

ASK CLI starts a local HTTP server listening on http://localhost:9090/
ASK CLI generates an Amazon sign-in URL with the parameter redirect_uri=http://127.0.0.1:9090/cb, then opens this URL in the user's web browser
The user signs in to Amazon in their web browser and enables ASK CLI to access their Amazon account
Amazon redirects the user's web browser back to http://127.0.0.1:9090/cb and passes the OAUTH auth code, scope, and state as GET parameters
ASK CLI issues a HTTP POST request to https://api.amazon.com/auth/o2/token to obtain an OAUTH access token and refresh token.
Another user on the same host can issue a HTTP request to http://127.0.0.1:9090/ and cause the ASK CLI process to terminate and cease listening on TCP port 9090. The attacker can then start their own server on TCP port 9090 and capture the OAUTH auth parameters, enabling them to hijack the OAUTH access token and refresh token. These can then be used to gain unauthorised access to the targeted user's Amazon account.

Stolen credentials could then in turn potentially be used to target the user more directly by injecting malicious code into the Alexa Skills that the user is working on.
Reproduction Steps

In a multi-user environment with at least two users:

As user A: Install ASK CLI and run the "ask configure" command, then wait for the console message Listening on http://localhost:9090... and the web browser to open
As user B: Issue a HTTP request to http://localhost:9090/ (e.g. using curl, wget, or a web browser)
As user B: Start a listener on TCP port 9090 (e.g. using nc -lvv 9090)
As user A: Proceed to sign in to Amazon and approve access for ASK CLI.
As user A: Close the browser and observe that the ASK CLI command has exited with no errors.
As user B: Observe that the OAUTH authentication parameters were received by the listener.

### PR DESCRIPTION
**Pull request by LucioMS**
Added a confirmation the browser has been used to sign in to alexa before closing the port.

DX for no:
```
% ask configure
This command will configure the ASK CLI with a profile associated with your Amazon developer credentials.
------------------------- Step 1 of 2 : ASK CLI Configuration -------------------------
? Please create a new profile or overwrite the existing profile.
 [x]                       ** NULL **
[Warn]: ASK CLI uses authorization code to fetch LWA tokens. Do not share neither your authorization code nor access tokens.
Switch to "Login with Amazon" page and sign-in with your Amazon developer credentials.
If your browser did not open the page, try to run the command again with "--no-browser" option.

? Do you confirm using the browser to sign in to Alexa Skills Kit Tools? No
[Error]: Stopping configuration with unconfirmed browser sign in.
``` 

DX for yes:
```
 % ask configure
This command will configure the ASK CLI with a profile associated with your Amazon developer credentials.
------------------------- Step 1 of 2 : ASK CLI Configuration -------------------------
? Please create a new profile or overwrite the existing profile.
 [x]                       ** NULL **
[Warn]: ASK CLI uses authorization code to fetch LWA tokens. Do not share neither your authorization code nor access tokens.
Switch to "Login with Amazon" page and sign-in with your Amazon developer credentials.
If your browser did not open the page, try to run the command again with "--no-browser" option.

? Do you confirm using the browser to sign in to Alexa Skills Kit Tools? Yes
ASK Profile "x" was successfully created. The details are recorded in ask-cli config file (.ask/cli_config) located at your **HOME** folder.
Vendor ID set as M1UCNFO4KV3AAA.

------------------------- Step 2 of 2 : Associate an AWS Profile with ASK CLI -------------------------
...
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: lib/controllers/authorization-controller/questions.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+module.exports = {
+  CONFIRM_ALLOW_BROWSER_SIGN_IN: [
+    {
+      message: "Do you confirm using the browser to sign in to Alexa Skills Kit Tools?",
```
Comment (tydonelson) [inline lib/controllers/authorization-controller/questions.js]: There isn't a clear indicator for the input of this prompt, and the grammar is slightly off. A couple of alternative suggestions:

'Do you confirm that you used the browser to sign in to Alexa Skills Kit Tools? [y/n]: '

'Press enter to continue when you have completed signing in with the browser... '

File: lib/controllers/authorization-controller/messages.js [hunk 2]
```diff
@@ -56,3 +56,5 @@ module.exports.ASK_SIGN_IN_FAILURE_MESSAGE = (error) => `
 `;
 
 module.exports.ASK_ENV_VARIABLES_ERROR_MESSAGE = "Could not find either of the environment variables: ASK_ACCESS_TOKEN, ASK_REFRESH_TOKEN";
+
+module.exports.STOP_UNCONFIRMED_BROWSER_SIGNIN = "Stopping configuration with unconfirmed browser sign in.";
```
Comment (tydonelson) [inline lib/controllers/authorization-controller/messages.js]: 'Stopping configuration due to unconfirmed browser sign in'

Comment (LucioMS) [inline lib/controllers/authorization-controller/messages.js]: Stopping configuration due to unconfirmed browser sign in.

**Review by tydonelson — APPROVED** [review]

File: lib/controllers/authorization-controller/index.js [hunk 3]
```diff
@@ -147,11 +148,29 @@ module.exports = class AuthorizationController {
...
       const requestQuery = url.parse(requestUrl, true).query;
-      server.destroy();
+
       if (requestUrl.startsWith("/cb?code")) {
         response.end(messages.ASK_SIGN_IN_SUCCESS_MESSAGE);
-        callback(null, requestQuery.code);
+        ui.confirmAllowSignIn((error, confirmSignInChoice) => {
+          // After confirmed or not browser sign in, closes the socket/port 
+          // with server.destroy().
+          // We need to keep the port open so a local hacker is not be able to
```
Comment (doiron) [inline lib/controllers/authorization-controller/index.js:157]: can we remove the part about local hackers

**Pull request merged**